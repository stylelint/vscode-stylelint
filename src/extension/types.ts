import type TypedEventEmitter from 'typed-emitter';
import type { DidRegisterDocumentFormattingEditProviderNotificationParams } from '../server/index';

/**
 * VS Code extension event names.
 */
export type ExtensionEvents = {
	DidRegisterDocumentFormattingEditProvider: (
		params: DidRegisterDocumentFormattingEditProviderNotificationParams,
	) => void;
	DidResetConfiguration: () => void;
};

/**
 * VS Code extension public API.
 */
export type PublicApi = TypedEventEmitter<ExtensionEvents> & {
	codeActionReady: boolean;
};

/**
 * Extension API event names.
 */
export enum ApiEvent {
	DidRegisterDocumentFormattingEditProvider = 'DidRegisterDocumentFormattingEditProvider',
	DidResetConfiguration = 'DidResetConfiguration',
}
