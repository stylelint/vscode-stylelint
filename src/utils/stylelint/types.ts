import type LSP from 'vscode-languageserver-protocol';
// eslint-disable-next-line n/no-unpublished-import
import type stylelint from 'stylelint';
import type { PackageManager } from '../packages/index';
export type Stylelint = typeof stylelint;
export type ConfigurationError = Error & { code: 78 };

/**
 * Diagnostics for a lint run.
 */
export type LintDiagnostics = {
	/**
	 * The diagnostics, each corresponding to a warning or error emitted by
	 * Stylelint.
	 */
	diagnostics: LSP.Diagnostic[];

	/**
	 * Raw output from Stylelint, if any.
	 */
	output?: string;

	/**
	 * Gets the original warning from the given diagnostic.
	 */
	getWarning?: (diagnostic: LSP.Diagnostic) => stylelint.Warning | null;
};

/**
 * Disable report rule names.
 */
export enum DisableReportRuleNames {
	Needless = '--report-needless-disables',
	InvalidScope = '--report-invalid-scope-disables',
	Descriptionless = '--report-descriptionless-disables',
	Illegal = 'reportDisables',
}

/**
 * Stylelint runner options.
 */
export type RunnerOptions = {
	config?: stylelint.Config | null;
	configBasedir?: string;
	configFile?: string;
	customSyntax?: string;
	ignoreDisables?: boolean;
	packageManager?: PackageManager;
	reportDescriptionlessDisables?: boolean;
	reportInvalidScopeDisables?: boolean;
	reportNeedlessDisables?: boolean;
	snippet?: string[];
	stylelintPath?: string;
	validate?: string[];
};

/**
 * Error thrown when a rule's option is invalid.
 */
export class InvalidOptionError extends Error {
	reasons: string[];

	constructor(warnings: { text: string }[]) {
		const reasons = warnings.map((warning) => warning.text);

		super(reasons.join('\n'));
		this.reasons = reasons;
	}
}
