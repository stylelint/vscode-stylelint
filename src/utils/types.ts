/**
 * Makes all properties of type `T` optional except for those in `K`.
 */
export type OptionalExcept<T extends Record<never, never>, K extends keyof T> = Pick<T, K> &
	Partial<Pick<T, K>>;

/**
 * Makes all properties of type `T` required except for those in `K`.
 */
export type RequiredExcept<T extends Record<never, never>, K extends keyof T> = Pick<
	T,
	Exclude<keyof T, K>
> &
	Required<Pick<T, K>>;

/**
 * Extracts keys from a type `T` the values of which are type `V`.
 */
export type ExtractKeysOfValueType<T, V> = {
	[K in keyof T]-?: T[K] extends V ? K : never;
}[keyof T];

/**
 * Given a function `T`, returns a function type with the same parameters, but
 * with the return type `Promise<R> | R` where `R` is the return type of `T`.
 */
// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type MaybeAsync<T extends (...args: any[]) => any> = (
	...args: Parameters<T>
) => Promise<ReturnType<T>> | ReturnType<T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T`.
 */
export type PublicOnly<T> = Pick<T, keyof T>;

/**
 * Given a type `T`, returns a type with only the public properties of `T` and
 * its descendants.
 */
export type PublicOnlyDeep<T> = {
	// eslint-disable-next-line @typescript-eslint/no-unsafe-function-type
	[K in keyof T]: T[K] extends Function ? T[K] : T[K] extends object ? PublicOnlyDeep<T[K]> : T[K];
};
