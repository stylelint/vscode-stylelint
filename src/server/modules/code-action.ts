import type { TextDocument } from 'vscode-languageserver-textdocument';
import * as LSP from 'vscode-languageserver-protocol';
import type winston from 'winston';
import { CodeActionKind as StylelintCodeActionKind, CommandId, Notification } from '../types';
import {
	RuleCodeActionsCollection,
	createDisableRuleFileCodeAction,
	createDisableRuleLineCodeAction,
} from '../../utils/lsp/index';
import { isDisableReportRule } from '../../utils/stylelint/index';
import type {
	LanguageServerContext,
	LanguageServerModuleConstructorParameters,
	LanguageServerModule,
} from '../types';

export class CodeActionModule implements LanguageServerModule {
	static id = 'code-action';

	/**
	 * The language server context.
	 */
	#context: LanguageServerContext;

	/**
	 * The logger to use, if any.
	 */
	#logger: winston.Logger | undefined;

	/**
	 * Disposables for notification and command handlers.
	 */
	#disposables: LSP.Disposable[] = [];

	constructor({ context, logger }: LanguageServerModuleConstructorParameters) {
		this.#context = context;
		this.#logger = logger;
	}

	async #shouldCodeAction(document: TextDocument): Promise<boolean> {
		const options = await this.#context.getOptions(document.uri);

		return options.validate.includes(document.languageId);
	}

	dispose(): void {
		this.#disposables.forEach((disposable) => disposable.dispose());
		this.#disposables.length = 0;
		this.#context.connection.onCodeAction(() => undefined);
	}

	onInitialize(): Partial<LSP.InitializeResult> {
		return {
			capabilities: {
				codeActionProvider: {
					codeActionKinds: [
						LSP.CodeActionKind.QuickFix,
						StylelintCodeActionKind.StylelintSourceFixAll,
					],
				},
				executeCommandProvider: {
					commands: [CommandId.OpenRuleDoc],
				},
			},
		};
	}

	onDidRegisterHandlers(): void {
		this.#logger?.debug('Registering onCodeAction handler');

		this.#context.connection.onCodeAction(async ({ context, textDocument: { uri } }) => {
			this.#logger?.debug('Received onCodeAction', { context, uri });

			const document = this.#context.documents.get(uri);

			if (!document) {
				this.#logger?.debug('Unknown document, ignoring', { uri });

				return [];
			}

			if (!(await this.#shouldCodeAction(document))) {
				this.#logger?.debug('Document should not be validated, ignoring', {
					uri,
					language: document.languageId,
				});

				return [];
			}

			const actions = await this.#getCodeActions(document, context);

			this.#logger?.debug('Returning code actions', { actions });

			return actions;
		});

		this.#logger?.debug('onCodeAction handler registered');

		this.#disposables.push(
			this.#context.commands.on(CommandId.OpenRuleDoc, async ({ arguments: args }) => {
				const params = args?.[0] as { uri: string } | undefined;

				if (!params) {
					this.#logger?.debug('No URL provided, ignoring command request');

					return {};
				}

				const { uri } = params;

				this.#logger?.debug('Opening rule documentation', { uri });

				// Open URL in browser
				const showURIResponse = await this.#context.connection.window.showDocument({
					uri,
					external: true,
				});

				if (!showURIResponse.success) {
					this.#logger?.warn('Failed to open rule documentation', { uri });

					return new LSP.ResponseError(
						LSP.ErrorCodes.InternalError,
						'Failed to open rule documentation',
					);
				}

				return {};
			}),
		);

		this.#disposables.push(
			this.#context.notifications.on(LSP.InitializedNotification.type, () =>
				this.#context.connection.sendNotification(Notification.DidRegisterCodeActionRequestHandler),
			),
		);
	}

	async #getAutoFixAllAction(document: TextDocument): Promise<LSP.CodeAction | undefined> {
		const edits = await this.#context.getFixes(document);

		const identifier = { uri: document.uri, version: document.version };

		return edits.length > 0
			? LSP.CodeAction.create(
					'Fix all Stylelint auto-fixable problems',
					{ documentChanges: [LSP.TextDocumentEdit.create(identifier, edits)] },
					StylelintCodeActionKind.StylelintSourceFixAll,
				)
			: undefined;
	}

	#getQuickFixActions(document: TextDocument, context: LSP.CodeActionContext): LSP.CodeAction[] {
		const identifier = { uri: document.uri, version: document.version };
		const actions: LSP.CodeAction[] = [];

		for (const diagnostic of context.diagnostics) {
			const edit = this.#context.getEditInfo(document, diagnostic);

			if (edit) {
				const action = LSP.CodeAction.create(
					edit.label,
					{ documentChanges: [LSP.TextDocumentEdit.create(identifier, [edit.edit])] },
					LSP.CodeActionKind.QuickFix,
				);

				actions.push(action);
			}
		}

		return actions;
	}

	#getAutoFixAllCommandAction(document: TextDocument): LSP.CodeAction {
		const command = LSP.Command.create(
			'Fix all Stylelint auto-fixable problems',
			CommandId.ApplyAutoFix,
			{ uri: document.uri, version: document.version },
		);

		return LSP.CodeAction.create(
			'Fix all Stylelint auto-fixable problems',
			command,
			LSP.CodeActionKind.Source,
		);
	}

	#getOpenRuleDocAction({ code, codeDescription }: LSP.Diagnostic): LSP.CodeAction | undefined {
		const uri = codeDescription?.href;

		if (!uri) {
			return undefined;
		}

		const command = LSP.Command.create(`Open documentation for ${code}`, CommandId.OpenRuleDoc, {
			uri,
		});

		return LSP.CodeAction.create(
			`Show documentation for ${code}`,
			command,
			LSP.CodeActionKind.QuickFix,
		);
	}

	async #getCodeActions(
		document: TextDocument,
		context: LSP.CodeActionContext,
	): Promise<LSP.CodeAction[]> {
		const only = context.only && new Set(context.only);

		this.#logger?.debug('Creating code actions', { only: context.only });
		const fixAllActions = [];

		if (
			only?.has(LSP.CodeActionKind.SourceFixAll) ||
			only?.has(StylelintCodeActionKind.StylelintSourceFixAll)
		) {
			this.#logger?.debug('Creating "source-fix-all" code action');

			const action = await this.#getAutoFixAllAction(document);

			if (action) {
				fixAllActions.push(action);
			}
		} else {
			this.#logger?.debug(
				'No source-fix-all actions requested, skipping source-fix-all action creation',
			);
		}

		if (only?.has(LSP.CodeActionKind.Source)) {
			this.#logger?.debug('Creating "source" code action');

			fixAllActions.push(this.#getAutoFixAllCommandAction(document));
		} else {
			this.#logger?.debug('No source actions requested, skipping source action creation');
		}

		if (
			only &&
			!only.has(LSP.CodeActionKind.QuickFix) &&
			!only.has(LSP.CodeActionKind.SourceFixAll) &&
			!only.has(StylelintCodeActionKind.StylelintSourceFixAll) &&
			!only.has(LSP.CodeActionKind.Source)
		) {
			this.#logger?.debug('No quick fix actions requested, skipping quick fix action creation');

			return fixAllActions;
		}

		// Otherwise, provide specific actions for each problem.
		const ruleCodeActions = new RuleCodeActionsCollection();

		for (const diagnostic of context.diagnostics) {
			const { source, code } = diagnostic;

			// If the diagnostic is not from Stylelint, ignore it.
			if (source !== 'Stylelint' || typeof code !== 'string') {
				continue;
			}

			// If the diagnostic is for a disable report, don't create disable
			// rule actions. Creating disable rule actions for an invalid
			// disable wouldn't make any sense.
			if (!isDisableReportRule(code)) {
				const options = await this.#context.getOptions(document.uri);
				const { location } = options.codeAction.disableRuleComment;

				this.#logger?.debug('Creating disable rule for line code action', {
					rule: code,
					location,
				});

				ruleCodeActions.get(code).disableLine = createDisableRuleLineCodeAction(
					document,
					diagnostic,
					location,
				);

				if (!ruleCodeActions.get(code).disableFile) {
					this.#logger?.debug('Creating disable rule for file code action', {
						rule: code,
					});

					ruleCodeActions.get(code).disableFile = createDisableRuleFileCodeAction(
						document,
						diagnostic,
					);
				}
			}

			if (!ruleCodeActions.get(code).documentation) {
				this.#logger?.debug('Creating documentation code action', {
					rule: code,
				});

				ruleCodeActions.get(code).documentation = this.#getOpenRuleDocAction(diagnostic);
			}
		}

		return [...this.#getQuickFixActions(document, context), ...ruleCodeActions, ...fixAllActions];
	}
}
