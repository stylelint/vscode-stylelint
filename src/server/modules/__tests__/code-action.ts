import { TextDocument } from 'vscode-languageserver-textdocument';
import * as LSP from 'vscode-languageserver-protocol';
import { CodeActionKind as StylelintCodeActionKind } from '../../types';

import { CodeActionModule } from '../code-action';
import { CommandId, Notification } from '../../index';
import { WorkDoneProgressReporter } from 'vscode-languageserver';

const mockContext = serverMocks.getContext();
const mockLogger = serverMocks.getLogger();

describe('CodeActionModule', () => {
	beforeEach(() => {
		mockContext.__options.validate = [];
		jest.clearAllMocks();
	});

	test('should be constructable', () => {
		expect(() => new CodeActionModule({ context: mockContext.__typed() })).not.toThrow();
	});

	test('onInitialize should return results', () => {
		const module = new CodeActionModule({ context: mockContext.__typed() });

		expect(module.onInitialize()).toMatchSnapshot();
	});

	test('onDidRegisterHandlers should register a code action handler', () => {
		const module = new CodeActionModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		expect(mockContext.connection.onCodeAction).toHaveBeenCalledTimes(1);
		expect(mockContext.connection.onCodeAction).toHaveBeenCalledWith(expect.any(Function));
	});

	test('onDidRegisterHandlers should register a InitializedRequest handler', () => {
		const module = new CodeActionModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		expect(mockContext.notifications.on).toHaveBeenCalledWith(
			LSP.InitializedNotification.type,
			expect.any(Function),
		);
	});

	test('should send the DidRegisterCodeActionRequestHandler notification when InitializedNotification is received', async () => {
		const module = new CodeActionModule({ context: mockContext.__typed() });

		module.onDidRegisterHandlers();

		await mockContext.notifications.on.mock.calls[0][1]?.();

		expect(mockContext.connection.sendNotification).toHaveBeenCalledWith(
			Notification.DidRegisterCodeActionRequestHandler,
		);
	});

	test('with action kind Source, should create fix-all code actions', async () => {
		mockContext.documents.get.mockReturnValue(
			TextDocument.create('foo', 'bar', 1, 'line 1\nline 2'),
		);
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([
			LSP.TextEdit.insert(LSP.Position.create(0, 0), 'text'),
		]);

		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: { only: [LSP.CodeActionKind.Source], diagnostics: [] },
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
	});

	test('with action kind StylelintSourceFixAll, should create fix-all code actions', async () => {
		const document = TextDocument.create('foo', 'bar', 1, 'line 1\nline 2');

		mockContext.documents.get.mockReturnValue(document);
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([
			LSP.TextEdit.insert(LSP.Position.create(0, 0), 'text'),
		]);

		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: {
					only: [StylelintCodeActionKind.StylelintSourceFixAll],
					diagnostics: [],
				},
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).toHaveBeenCalledWith(document);
	});

	test('with action kind SourceFixAll, should create fix-all code actions', async () => {
		const document = TextDocument.create('foo', 'bar', 1, 'line 1\nline 2');

		mockContext.documents.get.mockReturnValue(document);
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([
			LSP.TextEdit.insert(LSP.Position.create(0, 0), 'text'),
		]);

		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: { only: [LSP.CodeActionKind.SourceFixAll], diagnostics: [] },
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).toHaveBeenCalledWith(document);
	});

	test('with action kind SourceFixAll but no fixes, should not create fix-all code actions', async () => {
		const document = TextDocument.create('foo', 'bar', 1, 'line 1\nline 2');

		mockContext.documents.get.mockReturnValue(document);
		mockContext.__options.validate = ['bar'];
		mockContext.getFixes.mockResolvedValue([]);

		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: { only: [LSP.CodeActionKind.SourceFixAll], diagnostics: [] },
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).toHaveBeenCalledWith(document);
	});

	test('with no action kind, should create actions for each Stylelint diagnostic', async () => {
		mockContext.documents.get.mockReturnValue(
			TextDocument.create('foo', 'bar', 1, 'line 1\nline 2'),
		);
		mockContext.__options.validate = ['bar'];
		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const context: LSP.CodeActionContext = {
			diagnostics: [
				{
					message: 'Message for rule 1',
					source: 'Stylelint',
					range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
					code: 'rule 1',
					severity: LSP.DiagnosticSeverity.Error,
					codeDescription: {
						href: 'https://stylelint.io/user-guide/rules/rule',
					},
				},
				{
					message: 'Message for rule 2',
					source: 'ESLint',
					range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
					code: 'rule 2',
					severity: LSP.DiagnosticSeverity.Warning,
					codeDescription: {
						href: 'https://eslint.org/docs/rules/rule',
					},
				},
				{
					message: 'Message for rule 3',
					source: 'Stylelint',
					range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
					code: 'rule 3',
					severity: LSP.DiagnosticSeverity.Warning,
				},
				{
					message: 'Message for rule 1',
					source: 'Stylelint',
					range: LSP.Range.create(LSP.Position.create(0, 1), LSP.Position.create(0, 1)),
					code: 'rule 1',
					severity: LSP.DiagnosticSeverity.Error,
					codeDescription: {
						href: 'https://stylelint.io/user-guide/rules/rule',
					},
				},
				{
					message: 'Message for rule 4',
					source: 'Stylelint',
					range: LSP.Range.create(LSP.Position.create(0, 1), LSP.Position.create(0, 1)),
					code: 404,
					severity: LSP.DiagnosticSeverity.Error,
				},
			],
		};

		const result = await handler(
			{
				context,
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 1)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toMatchSnapshot();
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Returning code actions', {
			actions: result,
		});
	});

	test('with unsupported action kind, should not create actions', async () => {
		mockContext.documents.get.mockReturnValue(
			TextDocument.create('foo', 'bar', 1, 'line 1\nline 2'),
		);
		mockContext.__options.validate = ['bar'];
		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: {
					only: ['foo'],
					diagnostics: [],
				},
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenCalledWith(
			'No quick fix actions requested, skipping action creation',
		);
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Returning code actions', { actions: [] });
	});

	test('if no matching document exists, should not attempt to create actions', async () => {
		mockContext.documents.get.mockReturnValue(undefined);

		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: { only: [LSP.CodeActionKind.Source], diagnostics: [] },
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith('Unknown document, ignoring', { uri: 'foo' });
	});

	test('if document language ID is not in options, should not attempt to create actions', async () => {
		mockContext.documents.get.mockReturnValue(
			TextDocument.create('foo', 'bar', 1, 'line 1\nline 2'),
		);
		mockContext.__options.validate = ['baz'];

		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();

		const handler = mockContext.connection.onCodeAction.mock.calls[0][0];

		const result = await handler(
			{
				context: { only: [LSP.CodeActionKind.Source], diagnostics: [] },
				textDocument: { uri: 'foo' },
				range: LSP.Range.create(LSP.Position.create(0, 0), LSP.Position.create(0, 0)),
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(result).toStrictEqual([]);
		expect(mockContext.getFixes).not.toHaveBeenCalled();
		expect(mockLogger.debug).toHaveBeenLastCalledWith(
			'Document should not be validated, ignoring',
			{
				uri: 'foo',
				language: 'bar',
			},
		);
	});

	describe('OpenRuleDoc command', () => {
		it('should open the rule documentation URL', async () => {
			mockContext.connection.window.showDocument.mockResolvedValue({
				success: true,
			});

			const module = new CodeActionModule({
				context: mockContext.__typed(),
				logger: mockLogger,
			});

			module.onDidRegisterHandlers();

			const handler = mockContext.commands.on.mock.calls.find(
				([command]) => command === CommandId.OpenRuleDoc,
			)?.[1];

			const result = await handler?.(
				{
					command: CommandId.OpenRuleDoc,
					arguments: [
						{
							uri: 'https://stylelint.io/user-guide/rules/foo',
						},
					],
				},
				{} as LSP.CancellationToken,
				{} as WorkDoneProgressReporter,
			);

			expect(result).toStrictEqual({});
			expect(mockContext.connection.window.showDocument).toHaveBeenCalledWith({
				uri: 'https://stylelint.io/user-guide/rules/foo',
				external: true,
			});
			expect(mockLogger.debug).toHaveBeenLastCalledWith('Opening rule documentation', {
				uri: 'https://stylelint.io/user-guide/rules/foo',
			});
			expect(mockLogger.warn).not.toHaveBeenCalled();
		});

		it('if no URL is provided, should do nothing', async () => {
			const module = new CodeActionModule({
				context: mockContext.__typed(),
				logger: mockLogger,
			});

			module.onDidRegisterHandlers();

			const handler = mockContext.commands.on.mock.calls.find(
				([command]) => command === CommandId.OpenRuleDoc,
			)?.[1];

			const result = await handler?.(
				{ command: CommandId.OpenRuleDoc },
				{} as LSP.CancellationToken,
				{} as WorkDoneProgressReporter,
			);

			expect(result).toStrictEqual({});
			expect(mockContext.connection.window.showDocument).not.toHaveBeenCalled();
			expect(mockLogger.debug).toHaveBeenLastCalledWith(
				'No URL provided, ignoring command request',
			);
			expect(mockLogger.warn).not.toHaveBeenCalled();
		});

		it('if opening the URL fails, should log a warning and respond with an error', async () => {
			mockContext.connection.window.showDocument.mockResolvedValue({
				success: false,
			});

			const module = new CodeActionModule({
				context: mockContext.__typed(),
				logger: mockLogger,
			});

			module.onDidRegisterHandlers();

			const handler = mockContext.commands.on.mock.calls.find(
				([command]) => command === CommandId.OpenRuleDoc,
			)?.[1];

			const result = await handler?.(
				{
					command: CommandId.OpenRuleDoc,
					arguments: [
						{
							uri: 'https://stylelint.io/user-guide/rules/foo',
						},
					],
				},
				{} as LSP.CancellationToken,
				{} as WorkDoneProgressReporter,
			);

			expect(result).toStrictEqual(
				new LSP.ResponseError(LSP.ErrorCodes.InternalError, 'Failed to open rule documentation'),
			);
			expect(mockContext.connection.window.showDocument).toHaveBeenCalledWith({
				uri: 'https://stylelint.io/user-guide/rules/foo',
				external: true,
			});
			expect(mockLogger.debug).toHaveBeenLastCalledWith('Opening rule documentation', {
				uri: 'https://stylelint.io/user-guide/rules/foo',
			});
			expect(mockLogger.warn).toHaveBeenLastCalledWith('Failed to open rule documentation', {
				uri: 'https://stylelint.io/user-guide/rules/foo',
			});
		});
	});

	it('should be disposable', () => {
		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		expect(module).toHaveProperty('dispose');
		expect(module.dispose).toBeInstanceOf(Function);
	});

	it('should set a no-op code action handler when disposed', async () => {
		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const handler = mockContext.connection.onCodeAction.mock.calls[1][0];

		const result = await handler(
			{
				context: {
					diagnostics: [],
					only: [LSP.CodeActionKind.QuickFix],
				},
				textDocument: { uri: 'foo' },
				range: {
					start: { line: 0, character: 0 },
					end: { line: 0, character: 0 },
				},
			},
			{} as LSP.CancellationToken,
			{} as WorkDoneProgressReporter,
		);

		expect(mockContext.connection.onCodeAction).toHaveBeenCalledTimes(2);
		expect(result).toBeUndefined();
	});

	it('should dispose all handler registrations when disposed', () => {
		const module = new CodeActionModule({
			context: mockContext.__typed(),
			logger: mockLogger,
		});

		module.onDidRegisterHandlers();
		module.dispose();

		const disposables = [
			...mockContext.notifications.on.mock.results,
			...mockContext.commands.on.mock.results,
		];

		expect(disposables).toHaveLength(2);
		expect(disposables[0].value.dispose).toHaveBeenCalledTimes(1);
		expect(disposables[1].value.dispose).toHaveBeenCalledTimes(1);
	});
});
