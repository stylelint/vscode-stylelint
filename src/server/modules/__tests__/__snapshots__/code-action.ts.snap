// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeActionModule onInitialize should return results 1`] = `
{
  "capabilities": {
    "codeActionProvider": {
      "codeActionKinds": [
        "quickfix",
        "source.fixAll.stylelint",
      ],
    },
    "executeCommandProvider": {
      "commands": [
        "stylelint.openRuleDoc",
      ],
    },
  },
}
`;

exports[`CodeActionModule with action kind Source, should create fix-all code actions 1`] = `
[
  {
    "command": {
      "arguments": [
        {
          "uri": "foo",
          "version": 1,
        },
      ],
      "command": "stylelint.applyAutoFix",
      "title": "Fix all Stylelint auto-fixable problems",
    },
    "kind": "source",
    "title": "Fix all Stylelint auto-fixable problems",
  },
]
`;

exports[`CodeActionModule with action kind SourceFixAll, should create fix-all code actions 1`] = `
[
  {
    "edit": {
      "documentChanges": [
        {
          "edits": [
            {
              "newText": "text",
              "range": {
                "end": {
                  "character": 0,
                  "line": 0,
                },
                "start": {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": {
            "uri": "foo",
            "version": 1,
          },
        },
      ],
    },
    "kind": "source.fixAll.stylelint",
    "title": "Fix all Stylelint auto-fixable problems",
  },
]
`;

exports[`CodeActionModule with action kind StylelintSourceFixAll, should create fix-all code actions 1`] = `
[
  {
    "edit": {
      "documentChanges": [
        {
          "edits": [
            {
              "newText": "text",
              "range": {
                "end": {
                  "character": 0,
                  "line": 0,
                },
                "start": {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": {
            "uri": "foo",
            "version": 1,
          },
        },
      ],
    },
    "kind": "source.fixAll.stylelint",
    "title": "Fix all Stylelint auto-fixable problems",
  },
]
`;

exports[`CodeActionModule with no action kind, should create actions for each Stylelint diagnostic 1`] = `
[
  {
    "edit": {
      "changeAnnotations": undefined,
      "documentChanges": [
        {
          "edits": [
            {
              "newText": "/* stylelint-disable-next-line rule 1 */
",
              "range": {
                "end": {
                  "character": 0,
                  "line": 0,
                },
                "start": {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": {
            "uri": "foo",
            "version": 1,
          },
        },
      ],
    },
    "kind": "quickfix",
    "title": "Disable rule 1 for this line",
  },
  {
    "edit": {
      "changeAnnotations": undefined,
      "documentChanges": [
        {
          "edits": [
            {
              "newText": "/* stylelint-disable rule 1 */
",
              "range": {
                "end": {
                  "character": 0,
                  "line": 0,
                },
                "start": {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": {
            "uri": "foo",
            "version": 1,
          },
        },
      ],
    },
    "kind": "quickfix",
    "title": "Disable rule 1 for the entire file",
  },
  {
    "command": {
      "arguments": [
        {
          "uri": "https://stylelint.io/user-guide/rules/rule",
        },
      ],
      "command": "stylelint.openRuleDoc",
      "title": "Open documentation for rule 1",
    },
    "kind": "quickfix",
    "title": "Show documentation for rule 1",
  },
  {
    "edit": {
      "changeAnnotations": undefined,
      "documentChanges": [
        {
          "edits": [
            {
              "newText": "/* stylelint-disable-next-line rule 3 */
",
              "range": {
                "end": {
                  "character": 0,
                  "line": 0,
                },
                "start": {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": {
            "uri": "foo",
            "version": 1,
          },
        },
      ],
    },
    "kind": "quickfix",
    "title": "Disable rule 3 for this line",
  },
  {
    "edit": {
      "changeAnnotations": undefined,
      "documentChanges": [
        {
          "edits": [
            {
              "newText": "/* stylelint-disable rule 3 */
",
              "range": {
                "end": {
                  "character": 0,
                  "line": 0,
                },
                "start": {
                  "character": 0,
                  "line": 0,
                },
              },
            },
          ],
          "textDocument": {
            "uri": "foo",
            "version": 1,
          },
        },
      ],
    },
    "kind": "quickfix",
    "title": "Disable rule 3 for the entire file",
  },
]
`;
